### PR TITLE
Decouple Poke Ball ids from item ids

### PR DESCRIPTION
**Pull request by AlexOn1ine**
Allows to add more pokeball without breaking saves. Makes use of `ItemIdToBallId` and the PokeBall enum. This also doesn't change as to how users have to add new pokeballs because everything was already present in expansion.

### CODE REVIEW

File: src/battle_anim_throw.c [hunk 1]
```diff
@@ -943,7 +941,7 @@ void AnimTask_IsBallBlockedByTrainer(u8 taskId)
     DestroyAnimVisualTask(taskId);
 }
 
-u8 ItemIdToBallId(u16 ballItem)
+enum PokeBall ItemIdToBallId(u16 ballItem)
```
Comment (Bassoonian) [inline src/battle_anim_throw.c:944]: I understand you may want to consider this out of scope, but imo we should really remove this function and define the `BALL_` ids in `items.h` in some field. This unification would be more in line with our current expansion approach.

Comment (AlexOn1ine) [reply src/battle_anim_throw.c:944]: You can't fully remove it because of the fail safe but I'm fine with this change

Comment (Bassoonian) [reply src/battle_anim_throw.c:944]: Then perhaps this function can just be "evaluate the field and provide the failsafe in case it's `NULL`"?

Comment (AlexOn1ine) [reply src/battle_anim_throw.c:944]: I could just keep the function and check the struct id instead. That would also reduce the number of changes I would need to make.

**Comment by AlexOn1ine** [comment]
ready for review 
pokeemerald-33

File: include/constants/items.h [hunk 2]
```diff
@@ -35,8 +35,8 @@
 // Note: If moving ball IDs around, updating FIRST_BALL/LAST_BALL is not sufficient
```
Comment (kittenchilly) [inline include/constants/items.h:35]: This note needs to be removed or changed.

Comment (AlexOn1ine) [reply include/constants/items.h]: I removed those defines since we can just use `POKEBALL_COUNT` now

File: data/scripts/debug.inc [hunk 3]
```diff
@@ -137,6 +137,7 @@ Debug_EventScript_Text_DefensiveIVs:
 	.string "HP IVs: {STR_VAR_1}, DEF IVs: {STR_VAR_2}, SPDEF IVs: {STR_VAR_3}$"
 
 Debug_EventScript_Script_1::
+	call MossdeepCity_SpaceCenter_2F_EventScript_ChoosePartyForMultiBattle
```
Comment (AsparagusEduardo) [inline data/scripts/debug.inc:140]:
```suggestion
```

File: src/battle_script_commands.c [hunk 4]
```diff
@@ -15175,16 +15175,16 @@ static void Cmd_handleballthrow(void)
         u32 odds, i;
         u8 catchRate;
 
-        gLastThrownBall = gLastUsedItem;
-        gBallToDisplay = gLastThrownBall;
+        gLastUsedItem = ItemIdToBallId(gLastUsedItem);
```
Comment (AsparagusEduardo) [inline src/battle_script_commands.c:15178]: I don't think it'd be a good idea to replace `gLastUsedItem` here, as there's the mismatch between Item IDs and ball IDs. On the lower sections, I would replace `gLastUsedItem` with `gLastThrownBall`.

File: src/battle_script_commands.c [hunk 5]
```diff
@@ -15364,8 +15364,8 @@ static void Cmd_handleballthrow(void)
         if (gBattleMons[gBattlerTarget].status1 & (STATUS1_POISON | STATUS1_BURN | STATUS1_PARALYSIS | STATUS1_TOXIC_POISON | STATUS1_FROSTBITE))
             odds = (odds * 15) / 10;
 
-        if (gBattleResults.catchAttempts[gLastUsedItem - FIRST_BALL] < 255)
-            gBattleResults.catchAttempts[gLastUsedItem - FIRST_BALL]++;
+        if (gBattleResults.catchAttempts[gLastUsedItem - BALL_POKE] < 255)
+            gBattleResults.catchAttempts[gLastUsedItem - BALL_POKE]++;
```
Comment (AsparagusEduardo) [inline src/battle_script_commands.c:15368]: This "-" assumes that the first ball is 1, when it's not anymore.

File: src/debug.c [hunk 6]
```diff
@@ -4174,12 +4174,12 @@ static void DebugAction_PCBag_Fill_PocketItems(u8 taskId)
 
 static void DebugAction_PCBag_Fill_PocketPokeBalls(u8 taskId)
 {
-    u16 itemId;
+    u16 ballId;
 
-    for (itemId = FIRST_BALL; itemId < LAST_BALL; itemId++)
+    for (ballId = BALL_POKE; ballId < POKEBALL_COUNT; ballId++)
```
Comment (AsparagusEduardo) [inline src/debug.c:4179]:
```suggestion
    for (ballId = BALL_STRANGE; ballId < POKEBALL_COUNT; ballId++)
```

File: src/event_object_movement.c [hunk 7]
```diff
@@ -7326,12 +7326,12 @@ static u8 LoadFillColorPalette(u16 color, u16 paletteTag, struct Sprite *sprite)
 static void ObjectEventSetPokeballGfx(struct ObjectEvent *objEvent)
 {
     #if OW_FOLLOWERS_POKEBALLS
-    u32 ball = BALL_POKE;
+    enum PokeBall ball = BALL_POKE;
```
Comment (AsparagusEduardo) [inline src/event_object_movement.c:7329]:
```suggestion
    enum PokeBall ball = BALL_STRANGE;
```

File: src/item.c [hunk 8]
```diff
@@ -166,11 +167,11 @@ bool8 HasAtLeastOneBerry(void)
 
 bool8 HasAtLeastOnePokeBall(void)
 {
-    u16 i;
+    u16 ballId;
 
-    for (i = FIRST_BALL; i <= LAST_BALL; i++)
+    for (ballId = BALL_POKE; ballId < POKEBALL_COUNT; ballId++)
```
Comment (AsparagusEduardo) [inline src/item.c:172]:
```suggestion
    for (ballId = BALL_STRANGE; ballId < POKEBALL_COUNT; ballId++)
```

File: src/pokeball.c [hunk 9]
```diff
@@ -47,36 +48,39 @@ static u16 GetBattlerPokeballItemId(u8 battlerId);
...
-#define GFX_TAG_FAST_BALL    55019
-#define GFX_TAG_HEAVY_BALL   55020
-#define GFX_TAG_DREAM_BALL   55021
-#define GFX_TAG_SAFARI_BALL  55022
-#define GFX_TAG_SPORT_BALL   55023
-#define GFX_TAG_PARK_BALL    55024
-#define GFX_TAG_BEAST_BALL   55025
-#define GFX_TAG_CHERISH_BALL 55026
+
+#define GFX_TAG_STRANGE_BALL 55000
```
Comment (AsparagusEduardo) [inline src/pokeball.c:52]:
```suggestion
#define GFX_TAG_STRANGE_BALL 55000
```

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]

File: src/battle_script_commands.c (hunk 4 above)
Comment (AlexOn1ine) [reply src/battle_script_commands.c:15178]: I made a local variable instead because it was scary to mutate both of `gLastThrownBall` and `gBallToDisplay`.

**Review by AsparagusEduardo — APPROVED** [review]

**Pull request merged**